src/core/ioc.py

```diff
@@ -0,0 +1,77 @@
+# structure for the ioc to have just one instance
+
+from src.core.singleton_decorator import singleton
+from src.lib.repositories.impl import *
+from src.api.controllers import *
+
+
+@singleton
+class Ioc:
+
+    _instance_ioc = {}
+
+    def __init__(self):
+
+        self._init_repositories()
+        self._init_controllers()
+
+    def _init_repositories(self):
+        self._instance_ioc[
+            "product_repository_impl"
+        ] = product_repository_impl.ProductRepositoryImpl()
+        self._instance_ioc[
+            "ingredient_repository_impl"
+        ] = ingredient_repository_impl.IngredientRepositoryImpl()
+        self._instance_ioc[
+            "inventory_ingredient_repository_impl"
+        ] = inventory_ingredient_repository_impl.InventoryIngredientRepositoryImpl()
+        self._instance_ioc[
+            "inventory_repository_impl"
+        ] = inventory_repository_impl.InventoryRepositoryImpl()
+        self._instance_ioc[
+            "order_repository_impl"
+        ] = order_repository_impl.OrderRepositoryImpl()
+        self._instance_ioc[
+            "order_detail_repository_impl"
+        ] = order_detail_repository_impl.OrderDetailRepositoryImpl()
+        self._instance_ioc[
+            "chef_repository_impl"
+        ] = chef_repository_impl.ChefRepositoryImpl()
+
+    def _init_controllers(self):
+        self._instance_ioc["product_controller"] = product_controller.ProductController(
+            self._instance_ioc["product_repository_impl"]
+        )
+        self._instance_ioc[
+            "ingredient_controller"
+        ] = ingredient_controller.IngredientController(
+            self._instance_ioc["ingredient_repository_impl"]
+        )
+        self._instance_ioc[
+            "inventory_ingredient_controller"
+        ] = inventory_ingredient_controller.InventoryIngredientController(
+            self._instance_ioc["inventory_ingredient_repository_impl"]
+        )
+        self._instance_ioc[
+            "inventory_controller"
+        ] = inventory_controller.InventoryController(
+            self._instance_ioc["inventory_repository_impl"]
+        )
+        self._instance_ioc["order_controller"] = order_controller.OrderController(
+            self._instance_ioc["order_repository_impl"]
+        )
+        self._instance_ioc[
+            "order_detail_controller"
+        ] = order_detail_controller.OrderDetailController(
+            self._instance_ioc["order_detail_repository_impl"]
+        )
+        self._instance_ioc["chef_controller"] = chef_controller.ChefController(
+            self._instance_ioc["chef_repository_impl"]
+        )
+
+    def put_instance(self, instance, instance_id):
```

Comment:
You should protect developers from override of your registered instance, either verify an existing instance can not be override it from the outside or hide `put_instance`

Reply:
done, i removed the put_instance method